apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,726 @@
 
-  /**
-   * Ensures that current panel is correctly set, if no current panel set and
-   * we're not in the process of transitioning then set default panel as current
-   * one.
-   */
-  ensureCurrentPanel: function() {
-    if (this.transitionPromise) {
-      return this.transitionPromise;
-    }
+  if (from) {
+    oldView = from.panel;
+    oldPanelElement = document.querySelector(`.panel-${oldView.name}`);
+  }
 
-    if (!currentPanel) {
-      return this.toPanel(this.defaultPanel);
-    }
+  newView = to.panel;
+  newPanelElement = document.querySelector(`.panel-${newView.name}`);
 
-    return Promise.resolve();
+  var doSlideAnimation = oldView && newView &&
+    oldPanelElement !== newPanelElement;
+  var isGoingBack = oldView && newView && oldView.previous === newView.name;
+
+  if (doSlideAnimation) {
+    newPanelElement.classList.add('panel-will-activate', 'panel-active');
+    oldPanelElement.classList.add('panel-will-deactivate');
+
+    newPanelElement.classList.toggle('panel-animation-back', isGoingBack);
+    oldPanelElement.classList.toggle('panel-animation-back', isGoingBack);
+
+    return waitForSlideAnimation(newPanelElement).then(() => {
+      oldPanelElement.classList.remove(
+        'panel-will-activate', 'panel-will-deactivate',
+        'panel-active', 'panel-animation-back'
+      );
+      newPanelElement.classList.remove(
+        'panel-will-deactivate', 'panel-will-activate', 'panel-animation-back'
+      );
+    }, () => {
+      oldPanelElement.classList.remove(
+        'panel-will-deactivate', 'panel-animation-back'
+      );
+      newPanelElement.classList.remove(
+        'panel-will-activate', 'panel-animation-back'
+      );
+    });
+  }
+
+  if (oldView) {
+    oldPanelElement.classList.remove('panel-active');
+  }
+
+  if (newView) {
+    newPanelElement.classList.add('panel-active');
+  }
+
+  return Promise.resolve();
+}
+
+/**
+ * Reacts to hash changes.
+ * 
+ * This is the heart of the navigation logic and handles running the various
+ * naviation steps.
+ * It also handles any unwanted errors. (TODO: needs to tune this a little
+ * more).
+ *
+ * @param {Event} e The hashchange event (not actually used in the function)
+ */
+function onHashChange(e) {
+  debug('onHashChange, location=', window.location);
+  var newPanel = findViewFromLocation(window.location);
+  if (!newPanel) {
+    return;
+  }
+
+  var args = Utils.params(decodeURIComponent(window.location.hash));
+
+  startNavigation({ newPanel, args }).then(
+    () => executeNavigationStep('beforeLeave')
+  ).then(
+    () => executeNavigationStep('beforeEnter').catch(catchStepError)
+  ).then(
+    switchPanel
+  ).then(
+    setFutureState
+  ).then(
+    () => executeNavigationStep('afterEnter').catch(catchStepError)
+  ).then(
+    () => executeNavigationStep('afterLeave').catch(catchStepError)
+  ).then(
+    endNavigation
+  ).catch(
+    onNavigationError // TODO reset to previous hash or call back ?
```

Comment:
I added this comment before adding "catch(catchStepError)" everywhere, all other functions are internal to the navigation and should not fail (right ? :) ). So maybe we can just do nothing